backends/dpdk/dpdkArch.cpp

```diff
@@ -772,26 +948,112 @@ const IR::Node *ConvertBinaryOperationTo2Params::postorder(IR::P4Parser *a) {
     return injector.inject_parser(parser, a);
 }
 
-const IR::Node *CollectLocalVariableToMetadata::preorder(IR::P4Program *p) {
+const IR::Node *CollectLocalVariables::preorder(IR::P4Program *p) {
     for (auto kv : structure->parsers) {
         locals_map.emplace(kv.first+"_parser", kv.second->parserLocals);
     }
     for (auto kv : structure->pipelines) {
         locals_map.emplace(kv.first, kv.second->controlLocals);
     }
     for (auto kv : structure->deparsers) {
-        locals_map.emplace(kv.first+"_deprser", kv.second->controlLocals);
+        locals_map.emplace(kv.first+"_deparser", kv.second->controlLocals);
     }
+    LOG4("Collecting local variables, locals_map:");
+    for (auto kv : locals_map) {
+        LOG4("  " << kv.first << ":");
+        for (auto decl : kv.second) {
+            LOG4("    " << decl);
+        }
+    }
+
+    /**
+     * It may happen that there are header definitions in the source P4 program
+     * which are placed after the definition of main header structure, which is
+     * perfectly fine if they are not used in the main header structure.
+     * But if such header is used to declare local variable there is created
+     * a new field of the same header type in the main header structure.
+     * If the main header structure is placed before the definition of that
+     * header, it is a problem for Type Checking.
+     * There might also be header definitions internally created by compiler
+     * which would cause the same problem.
+     * Therefore here we move all structure and header definitons directly
```

Comment:
I don't know if this is enough for the resulting code to always work. The headers may be used in other structs, not just in the main headers structure. In general, you need to insert the declarations before their first use and after the last declaration that they themselves use. 

But why can there be such headers? The headers you synthesize are not part of the main headers structure.

Reply:
This is only because of new headers created by ConvertLookahead pass which creates new header definitions in:
```
const IR::Node *ConvertLookahead::postorder(IR::P4Program *program) {
    auto origProgram = getOriginal()->to<IR::P4Program>();
    if (newHeaderMap.count(origProgram)) {
        IR::Vector<IR::Node>::iterator insertIter = std::find_if(program->objects.begin(),
                program->objects.end(), [](const IR::Node *n){return n->is<IR::P4Parser>();});
        auto newHeaders = newHeaderMap.at(origProgram);
        program->objects.insert(insertIter, newHeaders.begin(), newHeaders.end());

        LOG2("Following new headers inserted into program " << dbp(program) << ":");
        for (auto h : newHeaders) {
            LOG2(" " << h);
        }
    }
    return program;
}
```
and places them before first parser definition and it also creates new local variables of those header types. So after ConvertLookahead pass the locations of those header definitions are before they are first used. But in this pass, for each local variable which is a header (which applies also for new local variables created in ConverLookahead pass) there is created a corresponding field in the main headers structure, and therefore the definitions of those headers must be before the main headers structure. The headers which are synthesized in ConverLookahead pass are used only in the declarations of local variables (which are already after the definitions since the beginning when they are created) and in the main headers structure (there are created new fields of that header type during this pass), so that's why I move all the header and structure definitions before the main headers structure. This also moves other header and structure definitions which are not synthesized in ConvertLookahead pass, but the order of definitions is kept the same and they are only moved forward, so the only structure definition they get ahead is the main header structure and I assume that no header or structure definition can use the main header structure to define its fields.

Other approach could be inserting new header definitions synthesized in ConvertLookahead pass directly before main header structure at the moment when they are created and not before the first parser. That way this code would not be needed. That would probably be better approach. Currently the problem is that the name of the main headers structure is parsed in CollectMetadataHeaderInfo pass which is after the place where I placed ConvertLookahead pass, but probably ConvertLookahead pass could be placed after CollectMetadataHeaderInfo or the name of the main header structure could be parsed sooner.
I'm going to try this approach.

Reply:
This makes sense. However, if the code is never typechecked again after this pass you don't need the declarations in order really, you could simply generate code for them in the right order.

Reply:
But TypeChecking is called at least twice after this CollectLocalVariables pass.

Reply:
So do you think I should rather try this approach:

> Other approach could be inserting new header definitions synthesized in ConvertLookahead pass directly before main header structure at the moment when they are created and not before the first parser. That way this code would not be needed. That would probably be better approach. Currently the problem is that the name of the main headers structure is parsed in CollectMetadataHeaderInfo pass which is after the place where I placed ConvertLookahead pass, but probably ConvertLookahead pass could be placed after CollectMetadataHeaderInfo or the name of the main header structure could be parsed sooner.

or leave it as it is right now?

Reply:
Then probably it is best to insert everything there in `postorder(Type_Struct*)`. If the struct is the user metadata (which you have already identified, just check the type name) just return a Vector containing all new types inserted in front of the type_struct itself.

Reply:
BTW: this pattern is used several places in the compiler already, see for example ReplaceStructs.

Reply:
Or better `DoReplaceTuples::insertReplacements`

Reply:
Ok, will do. Thank you very much!

Reply:
done in 81951e53812cf830549b10fec008a371266659d1